pkg/config/validator.go

```diff
@@ -0,0 +1,236 @@
+				continue
+			}
+			aa.Params = acfg
+			p.validated.numAspects++
+			if validatePresence {
+				// ensure that aa.Kind has a registered adapter
+				if p.validated.adapterByKind[aa.GetKind()] == nil {
+					ce = ce.Appendf("Kind", "adapter for Kind=%s is not available", aa.GetKind())
+				}
+				if aa.GetAdapter() != "" && p.validated.adapterByName[aa.GetAdapter()] == nil {
+					ce = ce.Appendf("NamedAdapter", "adapter by name %s not available", aa.GetAdapter())
+				}
+			}
+		}
+		rs := rule.GetRules()
+		if len(rs) == 0 {
+			continue
+		}
+		if verr := p.validateAspectRules(rs, path, validatePresence); verr != nil {
+			ce = ce.Extend(verr)
+		}
+	}
+	return ce
+}
+
+// Validate validates a single serviceConfig and globalConfig together.
+// It returns a fully validated Config if no errors are found.
+func (p *Validator) Validate(serviceCfg string, globalCfg string) (rt *Validated, ce *aspect.ConfigErrors) {
+	var cerr *aspect.ConfigErrors
+	if re := p.validateGlobalConfig(globalCfg); re != nil {
+		cerr = ce.Appendf("GlobalConfig", "failed validation")
+		return rt, cerr.Extend(re)
+	}
+	// The order is important here, because serviceConfig refers to global config
+
+	if re := p.validateServiceConfig(serviceCfg, true); re != nil {
+		cerr = ce.Appendf("ServiceConfig", "failed validation")
+		return rt, cerr.Extend(re)
+	}
+
+	return p.validated, nil
+}
+
+// ValidateServiceConfig validates service config.
+// if validatePresence is true it will ensure that the named adapter and Kinds
+// have an available and configured adapter.
+func (p *Validator) validateServiceConfig(cfg string, validatePresence bool) (ce *aspect.ConfigErrors) {
+	var err error
+	m := &pb.ServiceConfig{}
+	if err = yaml.Unmarshal([]byte(cfg), m); err != nil {
+		ce = ce.Append("ServiceConfig", err)
+		return
+	}
+	if ce = p.validateAspectRules(m.GetRules(), "", validatePresence); ce != nil {
+		return ce
+	}
+	p.validated.serviceConfig = m
+	return
+}
+
+// UnknownValidator returns error for the given name.
+func UnknownValidator(name string) error {
+	return fmt.Errorf("unknown type [%s]", name)
+}
+
+// ConvertParams converts returns a typed proto message based on available Validator.
+func ConvertParams(finder ValidatorFinder, name string, params interface{}, strict bool) (proto.Message, error) {
+	var avl aspect.ConfigValidator
+	var found bool
+
+	if avl, found = finder.FindValidator(name); !found {
+		return nil, UnknownValidator(name)
+	}
+
+	acfg := avl.DefaultConfig()
+	if err := Decode(params, acfg, strict); err != nil {
+		return nil, err
+	}
+	if verr := avl.ValidateConfig(acfg); verr != nil {
+		return nil, verr
+	}
+	return acfg, nil
+}
+
+// Decode interprets src interface{} as the specified proto message.
+func Decode(src interface{}, dest proto.Message, strict bool) (err error) {
+	var md mapstructure.Metadata
+	var decoder *mapstructure.Decoder
+	if decoder, err = mapstructure.NewDecoder(&mapstructure.DecoderConfig{
```

Comment:
nit: build the DecoderConfig separately, or collapse the line to make it easier to parse the if clause.